### PR TITLE
Feature: Added api for getting/setting the complete state: rng, logits, embedding and kv_cache

### PR DESCRIPTION
**Pull request by xaedes**
I have implemented functions for getting and setting the rest of the model state.
It includes: random number generator state, logits, embedding and kv_cache.

It was necessary to store the logits so that we can `eval tokens`, `save state`, `restart program`, `load state` and then `sample`.
With just restoring kv_cache the sampling did not have access to the required logits and indeed segfaulted on the initially empty logits vector.

The logits vector initial capacity was reserved with a wrong value. This resulted in changing capacity after the first evaluation in which the logits vector is actually resized. I fixed this bug because it propagated to the state size, resulting in unnecessarily changes.

The random number generator state is also included to ensure consistent sampling results. 
Since the internal state of the rng is more than just the seed, it is serialized using the standard C++ api for this purpose by streaming into a stringbuffer. For simplicity I did not add further logic to parse and compress the serialized rng state.
For completeness I also stored the embedding vector.

Because the whole state is not in one contiguous memory buffer I decided on an output pointer parameter to get the state data.
The user is responsible to allocate the memory where the state is written to. To support this the required number of bytes can be requested.

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]
Great job!

File: llama.cpp [hunk 1]
```diff
@@ -2248,3 +2249,122 @@ const char * llama_print_system_info(void) {
 std::vector<std::pair<std::string, struct ggml_tensor *>>& llama_internal_get_tensor_map(struct llama_context * ctx) {
     return ctx->model.tensors_by_name;
 }
+
+// Returns the size of the state
+size_t llama_get_state_size(struct llama_context * ctx) {
+    const size_t s_bool = sizeof(int32_t);
```
Comment (ggerganov) [inline llama.cpp:2255]: `s_bool` is unused - is this expected?

Comment (xaedes) [reply llama.cpp:2255]: Oh, I missed that during cleanup of unused stuff.  
At one time during implementation I was saving the bool flags from llama_context, but removed it because it didnt make much sense.
